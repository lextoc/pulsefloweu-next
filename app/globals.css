.divider {
  margin: 3rem 0;
}

.content {
  margin-bottom: 25vh;
}

.cards {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(16rem, 20rem));
  gap: 1.5rem;
}
