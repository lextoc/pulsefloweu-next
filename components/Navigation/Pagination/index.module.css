.root {
  padding: 6rem 0;
}

.pagination {
  display: flex;
  align-items: center;
  justify-content: center;
}

.button {
  padding: 1rem;
  height: 3rem;
  width: 3rem;
  display: flex;
  align-items: center;
  justify-content: center;
  background: transparent;
  color: rgb(var(--secondary));
  cursor: pointer;
  font-weight: 700;
  transition: background 50ms ease;
  border: 0;
  border-bottom: 3px solid rgba(var(--secondary), 0.1);
}

.button:first-child {
  border-top-left-radius: var(--border-radius);
  border-bottom-left-radius: var(--border-radius);
}

.button:last-child {
  border-top-right-radius: var(--border-radius);
  border-bottom-right-radius: var(--border-radius);
}

.button:hover,
.button:focus {
  background: rgba(var(--secondary), 0.1);
}

.buttonActive {
  border-bottom: 3px solid rgb(var(--secondary));
}
