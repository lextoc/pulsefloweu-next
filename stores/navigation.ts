import { create } from "zustand";

interface NavigationData {
  menuTitle?: string;
  shouldToggleMobileMenu?: boolean;
  isMobileMenuOpen?: boolean;
}

interface NavigationState extends NavigationData {
  set: (data: Partial<NavigationData>) => void;
}

export const useNavigationStore = create<NavigationState>()((_set) => ({
  data: {
    menuTitle: "",
    shouldToggleMobileMenu: false,
    isMobileMenuOpen: false,
  },
  set: (data) =>
    _set(() => ({
      ...data,
    })),
}));
