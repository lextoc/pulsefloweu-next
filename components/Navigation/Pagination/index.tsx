import { usePathname, useRouter, useSearchParams } from "next/navigation";

import styles from "./index.module.css";

export interface PaginationProps {
  current_page: number;
  next_page: number | null;
  per_page: number;
  prev_page: number | null;
  total_pages: number;
  total_count: number;
}

export default function Pagination({
  current_page,
  next_page,
  per_page,
  prev_page,
  total_pages,
  total_count,
}: PaginationProps) {
  const router = useRouter();
  const pathname = usePathname();
  const searchParams = useSearchParams();

  const onClick = (page: number) => {
    const current = new URLSearchParams(Array.from(searchParams.entries()));
    if (current.delete) current.delete("page");
    current.set("page", `${page}`);
    router.push(`${pathname}?${current.toString()}`);
  };

  return (
    <div className={styles.root}>
      <div className={styles.pagination}>
        {Array.from(Array(total_pages), (e, index) => (
          <button
            key={index}
            className={`${styles.button} ${
              index + 1 === current_page ? styles.buttonActive : ""
            }`}
            onClick={() => onClick(index + 1)}
          >
            {index + 1}
          </button>
        ))}
      </div>
    </div>
  );
}
