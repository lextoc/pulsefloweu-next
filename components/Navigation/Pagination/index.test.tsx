import { expect } from "@jest/globals";
import { fireEvent, render } from "@testing-library/react";
import { usePathname, useRouter, useSearchParams } from "next/navigation";
import React from "react";

import Pagination, { PaginationProps } from "./index";

// Mock next/navigation module
jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
}));

describe("Pagination component", () => {
  const mockUseRouter = useRouter as jest.Mock;
  const mockUsePathname = usePathname as jest.Mock;
  const mockUseSearchParams = useSearchParams as jest.Mock;

  const mockRouterPush = jest.fn();

  beforeEach(() => {
    mockUseRouter.mockReturnValue({
      push: mockRouterPush,
    });
    mockUsePathname.mockReturnValue("/example-path");
    mockUseSearchParams.mockReturnValue(new URLSearchParams("page=1"));
  });

  const defaultProps: PaginationProps = {
    current_page: 1,
    next_page: 2,
    per_page: 10,
    prev_page: null,
    total_pages: 5,
    total_count: 40,
  };

  it("renders pagination buttons", () => {
    const { getByText } = render(<Pagination {...defaultProps} />);

    for (let page = 1; page <= 5; page++) {
      const button = getByText(page.toString());
      expect(button).toBeInTheDocument();
    }
  });

  it("applies active style to current page button", () => {
    const { getByText } = render(<Pagination {...defaultProps} />);

    const activeButton = getByText("1");
    expect(activeButton).toHaveClass("buttonActive");
  });

  it("calls router.push when clicking on a page button", () => {
    const { getByText } = render(<Pagination {...defaultProps} />);

    const button = getByText("2");
    fireEvent.click(button);

    expect(mockRouterPush).toHaveBeenCalledWith("/example-path?page=2");
  });

  it("does not render pagination when total_count is less than per_page", () => {
    const { container } = render(
      <Pagination {...defaultProps} total_count={5} />,
    );

    const pagination = container.querySelector(".pagination");

    expect(pagination).toBeNull();
  });

  it("does not render pagination when current_page is not provided", () => {
    const { container } = render(
      <Pagination {...defaultProps} current_page={undefined} />,
    );

    const pagination = container.querySelector(".pagination");

    expect(pagination).toBeNull();
  });
});
