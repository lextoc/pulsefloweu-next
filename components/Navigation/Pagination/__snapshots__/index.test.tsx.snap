// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component matches snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="pagination"
    >
      <button
        class="button buttonActive"
      >
        1
      </button>
      <button
        class="button "
      >
        2
      </button>
      <button
        class="button "
      >
        3
      </button>
      <button
        class="button "
      >
        4
      </button>
      <button
        class="button "
      >
        5
      </button>
    </div>
  </div>
</div>
`;
